src/policy/satisfy.rs

```diff
@@ -38,208 +39,135 @@ impl<'a, Pk: ToPublicKey> Satisfier<Pk> for PolicySatisfier<'a, Pk> {
+                super::serialize::and(&left, &right)
             }
-            Policy::Or { left, right } => {
-                // TODO: Replace std::u64::MAX in MSRV >=1.43.0
-                let mut left = left.satisfy_helper(satisfier).unwrap_or(SatisfierExtData {
-                    witness: VecDeque::new(),
-                    witness_cost: std::u64::MAX,
-                    program: left.compile(),
-                    program_cost: 0,
-                });
-                let mut right = right.satisfy_helper(satisfier).unwrap_or(SatisfierExtData {
-                    witness: VecDeque::new(),
-                    witness_cost: std::u64::MAX,
-                    program: right.compile(),
-                    program_cost: 0,
-                });
-
-                if left.cost() <= right.cost() {
-                    // Both left and right path are unsatisfiable
-                    // TODO: Replace std::u64::MAX in MSRV >=1.43.0
-                    if left.cost() == std::u64::MAX {
-                        return None;
+            Policy::Or {
+                ref left,
+                ref right,
+            } => {
+                let left = left.satisfy_internal(satisfier)?;
+                let right = right.satisfy_internal(satisfier)?;
+
+                // FIXME use a more intelligent cost function
+                let take_right = match (left.must_prune(), right.must_prune()) {
+                    (false, false) => {
+                        let left_cost = left.finalize()?.bounds().extra_cells;
+                        let right_cost = right.finalize()?.bounds().extra_cells;
+                        left_cost > right_cost
                     }
-
-                    left.witness.push_front(Value::u1(0));
-                    left.witness_cost += 1;
-                    left.program =
-                        compiler::or(left.program, right.program, UsedCaseBranch::Left).ok()?;
-                    left.program_cost += 6;
-                    Some(left)
+                    (false, true) => false,
+                    (true, false) => true,
+                    // If both children are must_prune the choice doesn't matter,
+                    // the case node itself will be marked must_prune.
+                    (true, true) => false,
+                };
+
+                if take_right {
+                    super::serialize::or(&left.pruned(), &right, Some(Arc::new(Value::u1(1))))
                 } else {
-                    right.witness.push_front(Value::u1(1));
-                    right.witness_cost += 1;
-                    right.program =
-                        compiler::or(left.program, right.program, UsedCaseBranch::Right).ok()?;
-                    right.program_cost += 6;
-                    Some(right)
+                    super::serialize::or(&left, &right.pruned(), Some(Arc::new(Value::u1(0))))
                 }
+                .prune_and_retype()
             }
```

Comment:
3030b553f76c1a51e79bd21cba68ed8e987b6565: The changing types after each pruning step make this very complicated: First we recursively satisfy the left and the right child. Then, in one `match` case, we finalize both which is another tree traversal. Combining left and right in an `or` is easy. Finally we reconstruct and type the resulting tree again.

That is the downside of the constructive approach. I'm not saying I have a better proposal right now... The fact that pruning causes non-local type changes forces us to reconstruct the entire tree from scratch again and again.

I hope that `WitnessNode::finalize()` is basically as fast as `Type::finalize()`.

Reply:
There are two "complicated" steps here, neither of which are really necessary:

* We call `.finalize` (which does indeed only finalize types, it doesn't do much other work) in order to get the cells bound. But in theory we could get this information directly from the `WitnessNode` without finalizng, we just haven't implemented that.
* We call `prune_and_retype` which is just an optimization at this point, to get us better estimates when making decisions in other `or`s and `thresh`s.